staticd/static_nb_config.c

```diff
@@ -0,0 +1,1150 @@
+
+	return NB_OK;
+}
+
+static int static_nexthop_bh_type_modify(struct nb_cb_modify_args *args)
+{
+	struct static_nexthop *nh;
+
+	nh = nb_running_get_entry(args->dnode, NULL, true);
+	nh->bh_type = yang_dnode_get_enum(args->dnode, NULL);
+
+	return NB_OK;
+}
+
+static int static_nexthop_bh_type_destroy(struct nb_cb_destroy_args *args)
+{
+	struct static_nexthop *nh;
+
+	nh = nb_running_get_entry(args->dnode, NULL, true);
+	nh->bh_type = STATIC_BLACKHOLE_NULL;
+
+	return NB_OK;
+}
+
+void routing_control_plane_protocols_control_plane_protocol_staticd_route_list_path_list_frr_nexthops_nexthop_apply_finish(
+	struct nb_cb_apply_finish_args *args)
+{
+	struct static_nexthop *nh;
+	struct static_path *pi;
+	struct route_node *rn;
+	const struct lyd_node *pi_dnode;
+	const struct lyd_node *rn_dnode;
+	const char *ifname;
+	const char *nh_vrf;
+	struct stable_info *info;
+	int nh_type;
+
+	nh_type = yang_dnode_get_enum(args->dnode, "./nh-type");
+	ifname = yang_dnode_get_string(args->dnode, "./interface");
+	nh_vrf = yang_dnode_get_string(args->dnode, "./vrf");
+
+	nh = nb_running_get_entry(args->dnode, NULL, true);
+
+	pi_dnode = yang_dnode_get_parent(args->dnode, "path-list");
+	pi = nb_running_get_entry(pi_dnode, NULL, true);
+
+	rn_dnode = yang_dnode_get_parent(pi_dnode, "route-list");
+	rn = nb_running_get_entry(rn_dnode, NULL, true);
+	info = route_table_get_info(rn->table);
+
+	static_install_nexthop(rn, pi, nh, info->safi, info->svrf, ifname,
+			       nh_type, nh_vrf);
+}
+
+
+void routing_control_plane_protocols_control_plane_protocol_staticd_route_list_src_list_path_list_frr_nexthops_nexthop_apply_finish(
+	struct nb_cb_apply_finish_args *args)
+{
+	struct static_nexthop *nh;
+	struct static_path *pi;
+	struct route_node *rn;
+	struct route_node *src_rn;
+	const struct lyd_node *pi_dnode;
+	const struct lyd_node *rn_dnode;
+	const struct lyd_node *src_dnode;
+	const char *ifname;
+	const char *nh_vrf;
+	struct stable_info *info;
+	int nh_type;
+
+	nh_type = yang_dnode_get_enum(args->dnode, "./nh-type");
+	ifname = yang_dnode_get_string(args->dnode, "./interface");
+	nh_vrf = yang_dnode_get_string(args->dnode, "./vrf");
+
+	nh = nb_running_get_entry(args->dnode, NULL, true);
+
+	pi_dnode = yang_dnode_get_parent(args->dnode, "path-list");
+	pi = nb_running_get_entry(pi_dnode, NULL, true);
+
+	src_dnode = yang_dnode_get_parent(pi_dnode, "src-list");
+	src_rn = nb_running_get_entry(src_dnode, NULL, true);
+
+	rn_dnode = yang_dnode_get_parent(src_dnode, "route-list");
+	rn = nb_running_get_entry(rn_dnode, NULL, true);
+	info = route_table_get_info(rn->table);
+
+	static_install_nexthop(src_rn, pi, nh, info->safi, info->svrf, ifname,
+			       nh_type, nh_vrf);
+}
+int routing_control_plane_protocols_control_plane_protocol_staticd_route_list_path_list_frr_nexthops_nexthop_mpls_label_stack_entry_pre_validate(
+	struct nb_cb_pre_validate_args *args)
+{
+	uint32_t pos;
+
+	pos = yang_get_list_pos(args->dnode);
+
+	if (pos > MPLS_MAX_LABELS) {
```

Comment:
This is relative expensive, since this callback is called for every label present in the nexthop label stack. It would be better to attach a pre_validate() callback to the nexthop list instead, and add a new helper function like `yang_dnode_count_elements()` to perform this check.

Reply:
Yes, It would be costly, shifted to nexthop.